stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1385 @@
+            query_rows(tx.tx(), qry, &[u64_to_sql(tenure_height)?])?;
+
+        if candidate_headers.len() == 0 {
+            // no nakamoto_block_headers at that tenure height, check if there's a stack block header where
+            //   block_height = tenure_height
+            let Some(ancestor_at_height) = tx
+                .get_ancestor_block_hash(tenure_height, tip_index_hash)?
+                .map(|ancestor| Self::get_block_header(tx.tx(), &ancestor))
+                .transpose()?
+                .flatten()
+            else {
+                return Ok(None);
+            };
+            // only return if it is an epoch-2 block, because that's
+            // the only case where block_height can be interpreted as
+            // tenure height.
+            if ancestor_at_height.is_epoch_2_block() {
+                return Ok(Some(ancestor_at_height));
+            } else {
+                return Ok(None);
+            }
+        }
+
+        for candidate in candidate_headers.into_iter() {
+            let Ok(Some(ancestor_at_height)) =
+                tx.get_ancestor_block_hash(tenure_height, tip_index_hash)
+            else {
+                // if there's an error or no result, this candidate doesn't match, so try next candidate
+                continue;
+            };
+            if ancestor_at_height == candidate.index_block_hash() {
+                return Ok(Some(candidate));
+            }
+        }
+        Ok(None)
+    }
+
+    /// Return the tenure height of `block` if it was a nakamoto block, or the
+    ///  Stacks block height of `block` if it was an epoch-2 block
+    ///
+    /// In Stacks 2.x, the tenure height and block height are the
+    /// same. A miner's tenure in Stacks 2.x is entirely encompassed
+    /// in the single Bitcoin-anchored Stacks block they produce, as
+    /// well as the microblock stream they append to it.
+    pub fn get_tenure_height(
+        conn: &Connection,
+        block: &StacksBlockId,
+    ) -> Result<Option<u64>, ChainstateError> {
+        let nak_qry = "SELECT tenure_height FROM nakamoto_block_headers WHERE index_block_hash = ?";
+        let opt_height: Option<i64> = conn
+            .query_row(nak_qry, &[block], |row| row.get(0))
+            .optional()?;
+        if let Some(height) = opt_height {
+            return Ok(Some(
+                u64::try_from(height).map_err(|_| DBError::ParseError)?,
+            ));
+        }
+
+        let epoch_2_qry = "SELECT block_height FROM block_headers WHERE index_block_hash = ?";
+        let opt_height: Option<i64> = conn
+            .query_row(epoch_2_qry, &[block], |row| row.get(0))
+            .optional()?;
+        opt_height
+            .map(u64::try_from)
+            .transpose()
+            .map_err(|_| ChainstateError::DBError(DBError::ParseError))
+    }
+
+    /// Load block header (either Epoch-2 rules or Nakamoto) by `index_block_hash`
+    pub fn get_block_header(
```

Comment:
Would it make sense to try the nakamoto block headers table first?  That might be more likely to return data.